cmd/service-controller/claim_verifier.go

```diff
@@ -0,0 +1,148 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"github.com/skupperproject/skupper/api/types"
+	"github.com/skupperproject/skupper/client"
+	"github.com/skupperproject/skupper/pkg/event"
+)
+
+const (
+	TokenClaimVerification string = "TokenClaimVerification"
+)
+
+type TokenGenerator interface {
+	ConnectorTokenCreate(ctx context.Context, subject string, namespace string) (*corev1.Secret, bool, error)
+}
+
+type ClaimVerifier struct {
+	vanClient *client.VanClient
+}
+
+func newClaimVerifier(client *client.VanClient) *ClaimVerifier {
+	return &ClaimVerifier{
+		vanClient: client,
+	}
+}
+
+func (server *ClaimVerifier) redeemClaim(name string, subject string, data []byte, generator TokenGenerator) (*corev1.Secret, string, int) {
+	claim, err := server.vanClient.KubeClient.CoreV1().Secrets(server.vanClient.Namespace).Get(name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil, "No such claim", http.StatusNotFound
+	} else if err != nil {
+		return nil, err.Error(), http.StatusInternalServerError
+	}
+	if claim.ObjectMeta.Labels == nil || claim.ObjectMeta.Labels[types.SkupperTypeQualifier] != types.TypeClaimRecord {
+		return nil, "No such claim", http.StatusNotFound
+	}
+	if claim.ObjectMeta.Annotations != nil {
+		if expirationString, ok := claim.ObjectMeta.Annotations[types.ClaimExpiration]; ok {
+			expiration, err := time.Parse(time.RFC3339, expirationString)
+			if err != nil {
+				event.Recordf(TokenClaimVerification, "Cannot determine expiration: %s", err)
+			} else if expiration.Before(time.Now()) {
+				event.Recordf(TokenClaimVerification, "Claim %s expired", name)
+				return nil, "No such claim", http.StatusNotFound
+			}
+		}
+	}
+	if bytes.Equal(claim.Data["password"], data) {
+		// generate a token and return it
+		token, _, err := generator.ConnectorTokenCreate(context.TODO(), subject, "")
+		if err != nil {
+			return nil, err.Error(), http.StatusInternalServerError
+		}
+		if claim.ObjectMeta.Annotations != nil {
+			if uses, ok := claim.ObjectMeta.Annotations[types.ClaimsRemaining]; ok {
+				remainingUses, err := strconv.Atoi(uses)
+				if err != nil {
+					event.Recordf(TokenClaimVerification, "Cannot determine remaining uses: %s", err)
```

Comment:
At present, the expectation is that if the claim record is there, it is still valid. When the claim is used up it is deleted. Errors deleting or updating the claim are considered internal issues and would not prevent the token being issued. I guess this last point is perhaps the question. If there is some issue updating the remaining uses (including deleting at 0), should that prevent the token being issued? It is I think a trade-off between strict enforcing of uses and user experience. I'm not sure which is better, wdyt?

Reply:
Actually, having thought about it, I agree that strict enforcing of the uses is the priority. I'll change this to retry to try and avoid spurious issues, but fail otherwise.

Reply:
Thanks. One last thing here. Should the `remainingUses` be updated inside the retry loop or before it?